### ISSUE TITLE
exp: empirical training validation for per-agent obs fix (#204) on COMPUTE_HOST_PRIMARY

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#220]
## Context

PR #216 (closes #204) landed the structural fix for the latent bug where `JointPPOTrainer.collect_rollout` was reusing one flat obs row across all N agents — so independent-PPO was provably feeding identical input to every policy.

The fix passes all unit tests (per-agent distinctness, identity-tail equals `np.eye(N)`, backward compat). But the open empirical question — **does fixing this bug measurably improve PPO learning?** — is unanswered. The Judge and Builder both explicitly deferred training validation to `COMPUTE_HOST_PRIMARY` per CLAUDE.md compute guidelines.

This issue tracks the empirical follow-up.

---

### Where the change lives (post-#216 surface)

- `bucket_brigade/training/joint_trainer.py` — `flatten_dict_obs(obs, agent_id=i, num_agents=N)` (line 58) appends a length-`N` identity one-hot; `JointPPOTrainer.collect_rollout` now builds per-agent rows and `RolloutBuffer.observations` is `[T, N, obs_dim]`. With `N=4` and the default 10-house layout, `obs_dim = 42 + 4 = 46`.
- `experiments/p3_specialization/train.py` — `CellConfig` driver; `obs_dim` is already probed with `agent_id=0, num_agents=cfg.num_agents` (line 450), so the per-cell training path is post-#216 wired correctly. No code change required to consume the fix.
- `experiments/p3_specialization/run_sweep.py` — preregistered grid driver. The `--scenarios`, `--lambdas`, `--seeds`, `--num-iterations`, `--rollout-steps` flags give us exactly the surface we need for a paired comparison.
- `experiments/p3_specialization/dropout_eval.py` — ablation pattern (single-agent dropout). Will be reused unchanged for the final treatment-cell evaluation; it already calls `flatten_dict_obs(..., agent_id=i, num_agents=...)` (line 77).
- `experiments/p3_specialization/diagnostics/inspect_rollout_rewards.py` — H1 script (per-agent CV, action-reward R²); use to measure per-agent policy divergence on the trained cells.
- `experiments/p3_specialization/diagnostics/issue199_baselines.py` — random + specialist baselines on `minimal_specialization` and `default`. Numbers from the 2026-05-15 notebook below are the pre-#216 reference points.

### Reference points (pre-#216, from research_notebook/2026-05-15_issue199_minimal_specialization.md)

Per-step team reward, 50 episodes, seed 42:

| Scenario | Random | Specialist | PPO iter-49 (pre-#216) | Gap closed |
|---|---|---|---|---|
| `minimal_specialization` | −96.07 | −22.07 | −75.80 | 27.4% (iter-49 spot); 18.1% (trailing-5) |
| `default` | 241.85 | 320.94 | (≈ random; see #199 / memory note) | ≤ 15% |

The treatment arm needs to close >50% of the specialist−random gap to flip the "PPO failure mode" memory note. The 18–27% pre-#216 number on `minimal_specialization` is the canonical baseline to beat.

### Experimental protocol

**Arms.** Paired comparison, same seeds, same config, only the codebase changes:
- **Baseline (pre-#216).** Check out `19afcd76` (the commit immediately before #216 / #207 — `feat(env): generalize Scenario ownership reward fields`). Verify `flatten_dict_obs` in `bucket_brigade/training/joint_trainer.py` does NOT take an `agent_id` arg.
- **Treatment (post-#216).** Current `main` (or any commit at-or-after #216 / `6d8bcc66`).

Use a separate worktree per arm on the remote so the two arms cannot collide.

**Scenarios.** `default` and `minimal_specialization`. `default` measures whether the fix closes the H2 plateau; `minimal_specialization` measures whether the fix closes the issue-#199 specialist gap (the cleaner test, per #199).

**Hyperparameters.** Match the issue #199 smoke configuration exactly so all numbers stack against the same reference:
- `--lambdas 0.0` (single arm; lambda_red sweep is orthogonal to this question).
- `--seeds 42 43 44` (3 seeds; sufficient to detect a 50% gap-closure effect at the observed cell-to-cell noise level).
- `--num-iterations 100` (twice the 50-iter smoke; required because the pre-#216 trajectory was still slope ≈ +0.029/iter — extending kills the "PPO just needed more time" objection; per #199 closing-paragraph language).
- `--rollout-steps 2048`
- `--num-agents 4`
- `--value-coef 0.5 --entropy-coef 0.01 --normalize-returns`
- `--device cpu` (the env is CPU-bound; per CLAUDE.md primary host guidance, GPU buys nothing).

**Cell count.** 2 arms × 2 scenarios × 3 seeds = **12 cells**.

**Commands.** Run on `COMPUTE_HOST_PRIMARY` in tmux:

```bash
source .env
HOST="$COMPUTE_HOST_PRIMARY"
ssh "$HOST"

# Treatment arm (current main, post-#216) ---------------------------------
cd ~/GitHub/bucket-brigade && git pull && git checkout main
tmux new -s p3_treat
uv sync
uv run python -m experiments.p3_specialization.run_sweep \
  --output-root experiments/p3_specialization/runs/issue220_treatment \
  --scenarios default minimal_specialization \
  --lambdas 0.0 \
  --seeds 42 43 44 \
  --num-iterations 100 \
  --rollout-steps 2048 \
  --num-agents 4 \
  --value-coef 0.5 --entropy-coef 0.01 --normalize-returns \
  --device cpu --skip-existing \
  2>&1 | tee experiments/p3_specialization/runs/issue220_treatment.log
# Ctrl-B D to detach.

# Baseline arm (revert #216 in a separate worktree) ----------------------
cd ~/GitHub/bucket-brigade
git fetch
git worktree add ../bb_pre216 19afcd76
cd ../bb_pre216
tmux new -s p3_base
uv sync
uv run python -m experiments.p3_specialization.run_sweep \
  --output-root experiments/p3_specialization/runs/issue220_baseline \
  --scenarios default minimal_specialization \
  --lambdas 0.0 \
  --seeds 42 43 44 \
  --num-iterations 100 \
  --rollout-steps 2048 \
  --num-agents 4 \
  --value-coef 0.5 --entropy-coef 0.01 --normalize-returns \
  --device cpu --skip-existing \
  2>&1 | tee experiments/p3_specialization/runs/issue220_baseline.log
# Ctrl-B D to detach.
```

**Wall-clock estimate.** Per `run_sweep.py`'s docstring (~3 min/cell on Mac Studio CPU for 50 iters × 2048 steps × hidden_size=64). Doubling iters → ~6 min/cell × 12 cells ≈ **75 min per arm sequentially** if run in one tmux. The two arms can run in parallel tmux sessions (the Mac Studio is comfortably multi-core), bringing wall-clock to ≈ 80 min total.

### Success metrics

Per cell, after the run finishes, harvest from `metrics.json`:

1. **Final per-step team reward** — mean of `mean_step_reward_team` over the last 5 iterations (matches the issue-#199 trailing-5 convention; iter-49 alone is too noisy).
2. **Per-agent action entropy** — `action_entropy/agent_{i}` for each `i`, plus `action_entropy/mean`. Identical-policy collapse (the pre-#216 failure mode) should manifest as `action_entropy` close to the baseline-arm mean while treatment-arm `action_entropy` per-agent diverges.
3. **Per-agent policy divergence (the structural metric this fix is supposed to enable).** Run the H1 inspector on each treatment cell to get per-agent CV and per-agent action-reward R²:
   ```bash
   uv run python experiments/p3_specialization/diagnostics/inspect_rollout_rewards.py \
     --cell experiments/p3_specialization/runs/issue220_treatment/<scenario>/lambda_0e0/seed_42 \
     --no-baseline
   ```
   Baseline arm should show ≈ identical per-agent CVs (the bug); treatment arm should show divergent CVs and per-agent R².
4. **Action-distribution KL between agent pairs** (new measurement — small script, ~30 LOC). For each cell, replay the final rollout, take each agent's per-step action distribution (`policy.get_action_distribution(obs[i])`), and compute pairwise KL averaged over steps. Baseline arm → KL ≈ 0; treatment arm → KL > 0 iff agents specialize.
5. **Gap-closed against #199 references.** For `minimal_specialization`, compute `(ppo_trailing5 − random) / (specialist − random)` against the 2026-05-15 baselines (random = −96.07, specialist = −22.07). The pre-#216 number was 18.1%. **Pass bar: treatment arm ≥ 50%** flips the "PPO failure mode" memory note; treatment arm < 25% reinforces the MAPPO recommendation.

### Where results go

- Cell outputs: `experiments/p3_specialization/runs/issue220_{baseline,treatment}/<scenario>/lambda_0e0/seed_<n>/` (already the layout that `run_sweep.py` produces; `metrics.json`, `config.json`, `policies/agent_*.pt`).
- Analysis script + tables: `experiments/p3_specialization/analyze_220.py` (model after `analyze_174.py` — small driver that walks both run roots, computes the 5 metrics above per cell, and emits a markdown table).
- Aggregated table + plots: `experiments/p3_specialization/diagnostics/results/issue220_obsfix/` (new directory; commit JSON + PNG artifacts there).
- Narrative writeup: `research_notebook/2026-05-16_issue220_obsfix_validation.md` (or current date when run). Follow the structure of `2026-05-15_issue199_minimal_specialization.md` — scenario table, metrics table, verdict against the >50% bar, recommended next step.

### Expected outcomes / interpretation

| Pre-registered outcome | Interpretation | Action |
|---|---|---|
| Treatment closes ≥ 50% of specialist−random gap on `minimal_specialization` | Obs-differentiation was the bottleneck; PPO is a viable algorithm. | Update "PPO failure mode" memory note to "fixed by #216"; downgrade #208 (MAPPO) priority. |
| Treatment closes 25–50% | Partial fix; some PPO failure remains. | Keep MAPPO follow-up active; document as joint-cause. |
| Treatment closes < 25% | Obs-differentiation wasn't the bottleneck. | "MAPPO is still needed" conclusion is firmer; document and prioritize #208. |

The per-agent action-entropy divergence and pairwise KL > 0 are *necessary* conditions for the fix to have worked at all — even if the team-reward number doesn't move, divergence > 0 confirms the structural bug is gone and points the next intervention at credit assignment rather than identical-gradient collapse.

---

## What to do (original)

Run a paired comparison on the primary compute host:

1. **Baseline arm:** revert PR #216 (or run a pre-#216 commit). Train independent-PPO on `default` and `minimal_specialization` for 50-100 iterations, 3+ seeds.
2. **Treatment arm:** current main (post-#216). Same training config, same seeds.
3. Compare:
   - Final per-step team reward (does the bug fix close the gap to random?)
   - Per-agent action diversity (the bug should have collapsed agents toward identical policies)
   - Wall-clock to match a fixed reward target

## References

- Issue #204 (curator analysis)
- PR #216 (the fix)
- PR #215 (env-health regression suite — shared `bucket_brigade/diagnostics/` helpers used by H1)
- Memory note: "PPO failure mode — env-side fixes verified working (#197, #198, #199); independent-PPO still plateaus at 15% of optimum on a sanity-check scenario." This is the experiment that tests whether obs-differentiation closes the gap.
- `research_notebook/2026-05-15_issue199_minimal_specialization.md` — reference baselines + verdict template.

## Acceptance criteria

- [ ] Training runs completed on COMPUTE_HOST_PRIMARY (12 cells total: 2 arms × 2 scenarios × 3 seeds × 100 iters), results committed under `experiments/p3_specialization/runs/issue220_{baseline,treatment}/`.
- [ ] `experiments/p3_specialization/analyze_220.py` (new) walks both run roots and emits the five metrics per cell as a markdown table; output saved under `experiments/p3_specialization/diagnostics/results/issue220_obsfix/`.
- [ ] Dated research_notebook entry (`research_notebook/2026-05-16_issue220_obsfix_validation.md` or current-date) summarizing the comparison, with the verdict mapped onto the pre-registered table above.
- [ ] H1 inspector run on each treatment cell; per-agent CV / R² recorded.
- [ ] If the fix closes ≥ 50% of the gap on `minimal_specialization`: project memory updated (the "PPO failure mode" memory) and #208 priority lowered. If 25–50%: documented joint-cause. If < 25%: memory tightened to "MAPPO is necessary; obs-differentiation alone insufficient" and #208 raised.
